File: kuma_sync/docker/sync.py
import subprocess
import yaml
import re
import requests


def send_gotify_alert(gotify_url, gotify_token, title, message, priority=5):
    """Send alert to Gotify"""
    try:
        resp = requests.post(
            f"{gotify_url}/message",
            data={"title": title, "message": message, "priority": priority},
            headers={"X-Token": gotify_token},
            timeout=10,
        )
        return resp.status_code == 200
    except Exception as e:
        print(f"[!] Failed to send Gotify alert: {e}")
        return False


class KumaAPI:
    """Direct REST API client for Uptime Kuma v2"""

    def __init__(self, url):
        self.url = url.rstrip("/")
        self.token = None
        self.session = requests.Session()

    def login(self, username, password):
        resp = self.session.post(
            f"{self.url}/api/login",
            json={"username": username, "password": password},
            timeout=30,
        )
        resp.raise_for_status()
        data = resp.json()
        self.token = data.get("token")
        if not self.token:
            raise Exception("No token received from login")
        return True

    def _headers(self):
        return {"Authorization": f"Bearer {self.token}"}

    def get_monitors(self):
        resp = self.session.get(
            f"{self.url}/api/monitors",
            headers=self._headers(),
            timeout=30,
        )
        resp.raise_for_status()
        return resp.json().get("monitors", [])

    def get_docker_hosts(self):
        resp = self.session.get(
            f"{self.url}/api/docker-hosts",
            headers=self._headers(),
            timeout=30,
        )
        resp.raise_for_status()
        return resp.json()

    def add_monitor(self, monitor_type, name, url=None, docker_container=None,
                   docker_host=None, interval=60, retry_interval=60, max_retries=3):
        payload = {
            "name": name,
            "type": monitor_type,
            "interval": interval,
            "retryInterval": retry_interval,
            "maxretries": max_retries,
            "conditions": [],
        }

        if monitor_type == "http":
            payload["url"] = url
            payload["method"] = "GET"
            payload["accepted_statuscodes"] = [200, 201, 204, 301, 302]
        elif monitor_type == "docker":
            payload["docker_container"] = docker_container
            payload["docker_host"] = docker_host

        resp = self.session.post(
            f"{self.url}/api/monitors",
            headers=self._headers(),
            json=payload,
            timeout=30,
        )
        resp.raise_for_status()
        return resp.json()


def parse_healthcheck(service_name, service_data):
    healthcheck = service_data.get("healthcheck", {})
    test = healthcheck.get("test", [])

    if isinstance(test, list):
        test_str = " ".join(test)
    else:
        test_str = str(test)

    match = re.search(r"https?://(?:localhost|127\.0\.0\.1)(?::(\d+))?(/[\w/.-]*)?", test_str)
    if match:
        port = match.group(1) or "80"
        path = match.group(2) or "/"

        network_mode = service_data.get("network_mode", "")
        if network_mode.startswith("service:"):
            hostname = network_mode.split(":")[1]
        else:
            hostname = service_data.get("container_name", service_name)

        return f"http://{hostname}:{port}{path}"
    return None


def get_services(compose_path):
    if not os.path.exists(compose_path):
        raise FileNotFoundError(f"docker-compose.yml not found: {compose_path}")

    with open(compose_path, "r") as f:
        compose_data = yaml.safe_load(f)

    return compose_data.get("services", {})


def _synccontainers_try_run(services, api, docker_host_id, dry_run=False, existing=None):
    added = 0
    skipped = 0
    new_monitors = []

    for service_name, service_data in services.items():
        display_name = service_data.get("container_name", service_name)

        if display_name in existing:
            print(f"[-] {display_name}: already exists, skipping")
            skipped += 1
            continue

        url = parse_healthcheck(service_name, service_data)

        if dry_run:
            if url:
                print(f"[+] {display_name}: would add HTTP monitor -> {url}")
            else:
                print(f"[+] {display_name}: would add Docker monitor")
            added += 1
            continue

        try:
            if url:
                print(f"[+] {display_name}: adding HTTP monitor -> {url}")
                api.add_monitor(
                    monitor_type="http",
                    name=display_name,
                    url=url,
                )
                new_monitors.append(display_name)
                added += 1
            else:
                container_id = service_data.get("container_name", service_name)
                print(f"[+] {display_name}: adding Docker monitor (container: {container_id})")
                api.add_monitor(
                    monitor_type="docker",
                    name=display_name,
                    docker_container=container_id,
                    docker_host=docker_host_id,
                )
                new_monitors.append(display_name)
                added += 1
        except Exception as e:
            print(f"[!] {display_name}: failed - {e}")

    return {"added": added, "skipped": skipped, "new_monitors": new_monitors}


def synccontainers(compose_path, kuma_url, kuma_user, kuma_pass, docker_host_name=None,
                 docker_socket="/var/run/docker.sock", dry_run=False,
                 gotify_url=None, gotify_token=None):
    import os

    services = get_services(compose_path)
    if not services:
        print("No services found in docker-compose.yml")
        return {"added": 0, "skipped": 0}

    print(f"Found {len(services)} services in {compose_path}")

    if dry_run:
        return _synccontainers_try_run(services, None, None, dry_run=True)

    api = KumaAPI(kuma_url)
    try:
        api.login(kuma_user, kuma_pass)
    except Exception as e:
        print(f"Login failed: {e}")
        return {"added": 0, "skipped": 0}

    docker_hosts = api.get_docker_hosts()
    docker_host_id = None
    for host in docker_hosts:
        if docker_host_name and host.get("name") == docker_host_name:
            docker_host_id = host.get("id")
            break
    if not docker_host_id and docker_hosts:
        docker_host_id = docker_hosts[0].get("id")
        print(f"Using Docker host: {docker_hosts[0].get('name')} (ID: {docker_host_id})")
    else:
        print("No Docker host found in Uptime Kuma")
        print("Please add a Docker host first via the Uptime Kuma UI")
        return {"added": 0, "skipped": 0}

    existing = {m["name"]: m for m in api.get_monitors()}
    print(f"Existing monitors: {len(existing)}")

    result = _synccontainers_try_run(services, api, docker_host_id, dry_run=False, existing=existing)

    if gotify_url and gotify_token and result["added"] > 0:
        msg = f"Added {result['added']} new monitors to Uptime Kuma:\n"
        msg += "\n".join(f"• {s}" for s in result["new_monitors"][:10])
        if len(result["new_monitors"]) > 10:
            msg += f"\n... and {len(result['new_monitors']) - 10} more"
        send_gotify_alert(gotify_url, gotify_token, "New Monitors Added", msg)

    return result

File: kuma_sync_go/docker/sync.go
package docker

import (
	"fmt"
	"os"
	"regexp"

	"gopkg.in/yaml.v3"

	"kuma-sync/kumasync"
)

type ServiceData struct {
	ContainerName string       `yaml:"container_name"`
	HealthCheck  HealthCheck  `yaml:"healthcheck"`
	NetworkMode string        `yaml:"network_mode"`
}

type HealthCheck struct {
	Test any `yaml:"test"`
}

type ComposeFile struct {
	Services map[string]ServiceData `yaml:"services"`
}

func ParseHealthcheck(serviceName string, serviceData ServiceData) string {
	healthcheck := serviceData.HealthCheck
	test := healthcheck.Test

	var testStr string
	switch v := test.(type) {
	case []any:
		testStr = ""
		for _, t := range v {
			testStr += fmt.Sprintf("%v ", t)
		}
		testStr = testStr[:len(testStr)-1]
	case string:
		testStr = v
	default:
		testStr = ""
	}

	re := regexp.MustCompile(`https?://(?:localhost|127\.0\.0\.1)(?::(\d+))?(/[\w/.-]*)?`)
	matches := re.FindStringSubmatch(testStr)
	if matches == nil {
		return ""
	}

	port := "80"
	if matches[1] != "" {
		port = matches[1]
	}
	path := "/"
	if matches[2] != "" {
		path = matches[2]
	}

	hostname := serviceData.ContainerName
	if serviceName == "" {
		hostname = serviceName
	}

	networkMode := serviceData.NetworkMode
	if len(networkMode) > 8 && networkMode[:8] == "service:" {
		hostname = networkMode[8:]
	} else if serviceData.ContainerName != "" {
		hostname = serviceData.ContainerName
	}

	return fmt.Sprintf("http://%s:%s%s", hostname, port, path)
}

func GetServices(composePath string) (map[string]ServiceData, error) {
	if _, err := os.Stat(composePath); os.IsNotExist(err) {
		return nil, fmt.Errorf("docker-compose.yml not found: %s", composePath)
	}

	data, err := os.ReadFile(composePath)
	if err != nil {
		return nil, err
	}

	var compose ComposeFile
	if err := yaml.Unmarshal(data, &compose); err != nil {
		return nil, err
	}

	return compose.Services, nil
}

func SyncContainers(composePath, kumaURL, kumaUser, kumaPass, dockerHostName string,
	dockerSocket string, dryRun bool, gotifyURL, gotifyToken string) kumasync.SyncResult {

	services, err := GetServices(composePath)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return kumasync.SyncResult{}
	}

	if len(services) == 0 {
		fmt.Println("No services found in docker-compose.yml")
		return kumasync.SyncResult{}
	}

	fmt.Printf("Found %d services in %s\n", len(services), composePath)

	if dryRun || kumaURL == "" {
		added := 0
		for serviceName, serviceData := range services {
			displayName := serviceData.ContainerName
			if displayName == "" {
				displayName = serviceName
			}
			url := ParseHealthcheck(serviceName, serviceData)
			if url != "" {
				fmt.Printf("[+] %s: would add HTTP monitor -> %s\n", displayName, url)
			} else {
				fmt.Printf("[+] %s: would add Docker monitor\n", displayName)
			}
			added++
		}
		return kumasync.SyncResult{Added: added}
	}

	api := kumasync.NewKumaAPI(kumaURL)
	if err := api.Login(kumaUser, kumaPass); err != nil {
		fmt.Printf("Login failed: %v\n", err)
		return kumasync.SyncResult{}
	}

	dockerHosts, err := api.GetDockerHosts()
	if err != nil {
		fmt.Printf("Failed to get docker hosts: %v\n", err)
		return kumasync.SyncResult{}
	}

	dockerHostID := 0
	if dockerHostName != "" {
		for _, host := range dockerHosts {
			if host.Name == dockerHostName {
				dockerHostID = host.ID
				break
			}
		}
	}

	if dockerHostID == 0 && len(dockerHosts) > 0 {
		dockerHostID = dockerHosts[0].ID
		fmt.Printf("Using Docker host: %s (ID: %d)\n", dockerHosts[0].Name, dockerHostID)
	} else if dockerHostID == 0 {
		fmt.Println("No Docker host found in Uptime Kuma")
		fmt.Println("Please add a Docker host first via the Uptime Kuma UI")
		return kumasync.SyncResult{}
	}

	existingMonitors, err := api.GetMonitors()
	if err != nil {
		fmt.Printf("Failed to get monitors: %v\n", err)
		return kumasync.SyncResult{}
	}

	existing := make(map[string]bool)
	for _, m := range existingMonitors {
		existing[m.Name] = true
	}
	fmt.Printf("Existing monitors: %d\n", len(existing))

	added := 0
	skipped := 0
	newMonitors := []string{}

	for serviceName, serviceData := range services {
		displayName := serviceData.ContainerName
		if displayName == "" {
			displayName = serviceName
		}

		if existing[displayName] {
			fmt.Printf("[-] %s: already exists, skipping\n", displayName)
			skipped++
			continue
		}

		url := ParseHealthcheck(serviceName, serviceData)

		tryAdd := true
		if dryRun {
			tryAdd = false
		}

		if tryAdd {
			if url != "" {
				fmt.Printf("[+] %s: adding HTTP monitor -> %s\n", displayName, url)
				api.AddMonitor("http", displayName, url, "", dockerHostID, 60, 60, 3)
				newMonitors = append(newMonitors, displayName)
				added++
			} else {
				containerID := serviceData.ContainerName
				if containerID == "" {
					containerID = serviceName
				}
				fmt.Printf("[+] %s: adding Docker monitor (container: %s)\n", displayName, containerID)
				api.AddMonitor("docker", displayName, "", containerID, dockerHostID, 60, 60, 3)
				newMonitors = append(newMonitors, displayName)
				added++
			}
		} else {
			if url != "" {
				fmt.Printf("[+] %s: would add HTTP monitor -> %s\n", displayName, url)
			} else {
				fmt.Printf("[+] %s: would add Docker monitor\n", displayName)
			}
			newMonitors = append(newMonitors, displayName)
			added++
		}
	}

	result := kumasync.SyncResult{
		Added:       added,
		Skipped:     skipped,
		NewMonitors: newMonitors,
	}

	if gotifyURL != "" && gotifyToken != "" && added > 0 {
		msg := fmt.Sprintf("Added %d new monitors to Uptime Kuma:\n", added)
		for i, m := range newMonitors {
			if i >= 10 {
				break
			}
			msg += fmt.Sprintf("• %s\n", m)
		}
		if len(newMonitors) > 10 {
			msg += fmt.Sprintf("... and %d more\n", len(newMonitors)-10)
		}
		gotify := kumasync.NewGotifyClient(gotifyURL, gotifyToken)
		gotify.SendAlert("New Monitors Added", msg, 5)
	}

	return result
}